plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,290 @@
+
+import { ClusterDetails, KubernetesClustersSupplier } from '../types/types';
+import { KubernetesAuthTranslatorGenerator } from '../kubernetes-auth-translator/KubernetesAuthTranslatorGenerator';
+
+const HEADER_KUBERNETES_CLUSTERS: string = 'X-Kubernetes-Clusters';
+
+const ERROR_BAD_REQUEST: number = 400;
+const ERROR_INTERNAL_SERVER: number = 500;
+
+const TRANSLATOR_TYPE_TOKEN: string = 'token';
+const TRANSLATOR_TYPE_NOOP: string = 'serviceAccount';
+
+export interface KubernetesProxyServices {
+  kcs: KubernetesClustersSupplier;
+}
+
+export interface KubernetesProxyResponse {
+  code: number;
+  data: any;
+  cluster?: string;
+}
+
+interface KubernetesProxyClusters {
+  [key: string]: string;
+}
+
+export class KubernetesProxy {
+  constructor(protected readonly logger: Logger) {}
+
+  public async handleProxyRequest(
+    services: KubernetesProxyServices,
+    req: Request,
+  ): Promise<any> {
+    const krc = this.getKubernetesRequestedClusters(req);
+    const details = await this.getClusterDetails(services.kcs, krc);
+
+    const responses = await Promise.all(
+      details.map(async d => {
+        const response = await this.makeRequestToCluster(d, req);
+        return response;
+      }),
+    );
+
+    const data: { [key: string]: any } = {};
+    const codes: number[] = [];
+
+    responses.forEach(kpr => {
+      if (kpr.cluster) {
+        data[kpr.cluster] = kpr.data;
+        codes.push(kpr.code);
+      }
+    });
+
+    const code = this.getBestResponseCode(codes);
+
+    const res: KubernetesProxyResponse = {
+      code,
+      data,
+    };
+
+    return res;
+  }
+
+  protected getKubernetesRequestedClusters(
+    req: Request,
+  ): KubernetesProxyClusters {
+    const encodedClusters: string =
+      req.header(HEADER_KUBERNETES_CLUSTERS) ?? '';
+
+    if (!encodedClusters) {
+      return {};
+    }
+
+    try {
+      const decodedClusters = Buffer.from(encodedClusters, 'base64').toString();
+      const clusters: KubernetesProxyClusters = JSON.parse(decodedClusters);
+      return clusters;
+    } catch (e: any) {
+      this.logger.error(
+        `error with encoded cluster header: ${JSON.stringify(e)}`,
+      );
+    }
+    return {};
+  }
+
+  protected async getClusterDetails(
+    clusterSupplier: KubernetesClustersSupplier,
+    krc: KubernetesProxyClusters,
+  ): Promise<ClusterDetails[]> {
+    const clusters = await clusterSupplier.getClusters();
+
+    const clusterNames = Object.keys(krc);
+
+    const clusterDetails = clusters.filter(c => clusterNames.includes(c.name));
+
+    const clusterDetailsAuth = await Promise.all(
+      clusterDetails.map(async c => {
+        const authMetadata: KubernetesRequestAuth = {
+          token: krc[c.name],
+        };
+        const cAuth = await this.configureAuth(c, authMetadata);
+        return cAuth;
+      }),
+    );
+
+    return clusterDetailsAuth;
+  }
+
+  protected getClusterIP(details: ClusterDetails): string {
```

Comment:
FWIW the `server` field on a cluster is a URI, not an IP -- it will have a scheme like `https://` or and it can include a DNS hostname rather an an IP (EKS clusters look like `https://IDENTIFIER.STUFF.REGION.eks.amazonaws.com`)